### PR TITLE
Retry on dynamic index creation when an OpenSearchException is thrown

### LINKED ISSUE #3393: [BUG] Insufficient permissions to AOS shuts down DataPrepper when a dynamic index pattern is used
**Describe the bug**
As part of the sink initialization process, DataPrepper will attempt to create static indices. This also serves as a permissions check for the DataPrepper -> AOS integration. However, if a dynamic index pattern is used, no indices are created during initialization and instead permissions gaps are exposed at runtime by shutting down the pipeline:

```
ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [***] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: org.opensearch.client.opensearch._types.OpenSearchException: Request failed: [security_exception] no permissions for [cluster:monitor/state] and User [name=***, backend_roles=[***], requestedTenant=null]
```

**To Reproduce**
Create a pipeline with an FGAC-enabled AOS sink and a dynamic index pattern. Don't provide sufficient permissions to the pipeline role to ingest data to AOS. Ingest data into the DataPrepper pipeline to cause a crash.

**Expected behavior**
Similar to the behavior of static indices, DataPrepper should avoid crashing when there is a permissions gap and instead loop while providing helpful messaging to the user so that the issue can be resolved.

### PR DESCRIPTION
**Pull request by engechas**
### Description
Adds infinite retry when an OpenSearchException is thrown during dynamic index creation. This aligns with the behavior of static index creation.
 
### Issues Resolved
#3393 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by dlvenable — APPROVED** [review]

**Review by graytaylor0 — COMMENTED** [review]
What if it’s not a retryable error? For example, what if the dynamic index name contained capitalization. Would we infinitely retry?

**Review by graytaylor0 — APPROVED** [review]

**Comment by engechas** [comment]
>What if it’s not a retryable error? For example, what if the dynamic index name contained capitalization. Would we infinitely retry?

Is that exposed as an OpenSearch exception? If so then yes it would spin and retry. I am not sure there is a simple way to get more granular. We could try to match the text of the exception maybe but I think we'd end up with more crashes for truly retryable issues. Open to ideas.

My thought is that any exception from opensearch is an external dependency failure that requires rectification from the user. I think we should avoid crashes from external dependency failures in general and instead try to provide wording on how it can be fixed inline so the customer can take action.

Let me know if you feel otherwise and I can work on granularity here

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java [hunk 1]
```diff
@@ -7,18 +7,28 @@
 
 import com.github.benmanes.caffeine.cache.Cache;
 import com.github.benmanes.caffeine.cache.Caffeine;
+import org.opensearch.client.opensearch._types.OpenSearchException;
 import org.opensearch.client.RestHighLevelClient;
 import org.opensearch.client.opensearch.OpenSearchClient;
 import org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSinkConfiguration;
+
+import com.google.common.cache.Cache;
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java:15]: Please update to use only Caffeine. See #3586

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java (hunk 1 above)
Comment (engechas) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java]: Good catch, fixed

**Review by dlvenable — APPROVED** [review]

**Review by graytaylor0 — APPROVED** [review]

**Pull request merged**